Docs/company certificate standard/update (#1)
nicoprow/catenax-ev.github.io: docs/standards/CX-0135-CompanyCertificateManagement/CX-0135-CompanyCertificateManagement.md

Reviewer: I would like to keep the attribute: `"senderFeedbackUrl": "https://domain.tld/path/to/api",` as **optional** for all three feedback notifications (Received, Accepted, Rejected).
This would avoid adjustments and retesting.

Reviewer 2: From my understanding there can't be a response to a feedback notification, therefore making the `senderFeedbackUrl` obsolete for these payloads.

Reviewer: CTPAT | ctpat ist missing.

This standard is crucial for data providers and consumers who want to exchange company certificates through the Catena-X data space. By complying with this standard, companies can ensure seamless certificate management, thereby streamlining their overall operations.

The following company certificate use cases are supported in this release, including new entries as indicated
The following company certificate use cases are supported in this release:

1. Data Provider wants to publish a certificate.
2. (*new*) Data Consumer wants to request a certificate from a specific Data Provider
3. (*new*) Data Consumer wants to notify a Data Provider of acceptance or rejection of a certificate published in #1 via a status message

> 4. (*new*) Data Provider wants to notify a Data Consumer of the availability of a new Asset

Reviewer 3: We have added this entry text 6 month ago to enable the notification of a certificate consumer by the certificate provider that a new certificate is available. As a resulting action the certificate consumer can "pull" the certificate. Until now the certificate consumer has to periodically "poll" for required certificates. 
This feature is needed for certificate consumer who request data via the "pull" mechanism.
I can NOT agree to remove this option!

Reviewer 3: In the last regular meeting we decided to describe the business benefit of the "Certificate Available" notification. Based on this description and the following discussion we will decide related the above change.
Today some certificate consumers  use the "pull" and others use the "push" mechanism. We have agreed that both mechanism are needed and should be supported.
Certificate consumers who use the "pull" mechanism have the problem, that if a certificate is not available on provider side, then the consumer has to "poll", that means he has to request the certificate periodically and this creates effort and costs.
This problem can be solved if the certificate provider notifies the consumer if a new certificate instance is available (e.g. the successor of an existing expired certificate). The notification should contain all the data that are required for the consumer to request a certificate from the provider via the "pull" mechanism. The consumer will be interested to use this notification as a trigger to request the certificate instance via the pull mechanism.
The business benefit for the provider (as a supplier) is, that he can fullfill his delivery contract obligations to deliver company certificates on time to the consumer (as a customer).
The business benefit for the consumer is, that he does not need to configure the EDC with the BPNL specific entries referencing the bilateral contracts. Bigger companies have several thousands of providers and therefore the configuration effort is significant.

This standard is crucial for data providers and consumers who want to exchange company certificates through the Catena-X data space. By complying with this standard, companies can ensure seamless certificate management, thereby streamlining their overall operations.

The following company certificate use cases are supported in this release:

1. Certificate Distributor wants to publish a certificate.
2. Certificate Receiver wants to request a certificate from a specific Certificate Distributor
3. Certificate Receiver wants to notify a Certificate Distributor of acceptance or rejection of a certificate published in #1 via a feedback message

> 4. Certificate Distributor wants to notify a Certificate Receiver of the availability of a new available certificate (a new dataspace offer)

Reviewer 2: How would this be done without the Availability notification?

Reviewer 4: Maybe I'm missing it, but I think that also wasn't in before our rework. I think we just didn't consider this while re-writing. Either we have to add the notification mechanism, or we remove 4.. Perhaps this is something for a next iteration of the standard. I'd propose I remove 4. now. I'll also update 1. to include the case for the Certificate Receiver that wants to discover a provided certificate.

> *This section and all its subsections are normative*

Today, Certificate Consumer do not have a way to request certificates from a Certificate Provider. Also, Data Certificate Provider have no visibility on the status of a published certificate beyond the technical delivery. Finally, Certificate Provider do not have a standard mechanism to send new certificates to the Certificate Consumer when they become available.

Use cases to provide certificates (from who initiates communication, negotiation and data transfer):

> - Certificate Consumer -> Certificate Provider : Company Certificate Request

Reviewer 2: nit: technically the Company Certificate Request itself does not provide certificates, just the information if one is available and ready to be pulled. A Request can also be declined, not resulting in a certificate exchange.

```suggestion
- Certificate Consumer -> Certificate Provider : Company Certificate Pull
```

However, the pulling of a certificate can be done by itself without the need of a Request-notification.

Author: Fair enough I took over this change

- Certificate Provider -> Certificate Consumer : Company Certificate Push

Use case to give feedback for provided certificates:

- Certificate Consumer -> Certificate Provider : Company Certificate Status (Accepted, Rejected or Received)

### 2.1 API Specification

This section introduces the certificate management notification API which is further detailed in the corresponding [OpenAPI specification](assets/openapi-spec.yaml).

#### 2.1.1 API endpoints and resources

##### 2.1.1.1 Company Certificate Request

The Certificate Consumer is requesting a specific certificate from the Certificate Provider.
This request can be sent by the Certificate Consumer continuously, for updates on the request state on Certificate Provider side.

> ![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")

Reviewer 2: The descriptions of the HTTP response codes in the table from `2.1.1.1.1 HTTP Response Codes` and the Status Codes in the state machine do not align ("Request IN PROGRESS" on the bottom left in the state machine is a 200 while the table states that the 200 is only returned for a completed request).

Author: True, I changed it to 202 in the figure

> ![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")

Reviewer 2: The state transition from "Request is new & accepted: Status Code 202" points at the wrong state in my opinion - as this is returning a status code, it should point to the Start State on the (currently) bottom left of the graph because from there on a new request to identify the current status can be initiated.

Author: From a very strict point of view changing the transition is also not entirely correct, as in each state the process could undergo state transitions without the certificate consumer making a request. I think the idea of the state machine is clear but maybe needs to be a bit strictor formalized @FEILSTE ?

> ![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")

Reviewer 2: Identification of a Request-notification based on basically all of its attributes (senderBpn, receiverBpn, certifiedBpn, certificateType, locationBpns) is not ideal performance-wise, a single notificationId would be better.

Author: I agree but this creates a more complex interaction where an additional endpoint is needed. For the sake of not introducing many changes we kept it to these combined values instead to match certificate requests.

> *This section and all its subsections are normative*

Today, Data Consumer do not have a way to request certificates from a Data Provider. Also, Data Providers have no visibility on the status of a published certificate beyond the technical delivery. Finally, Data Providers do not have a standard mechanism to send new certificates to the Consumer when they become available.
We are adding the following 4 Application Programming Interfaces to remediate these shortcommings:
Today, Certificate Consumer do not have a way to request certificates from a Certificate Provider. Also, Data Certificate Provider have no visibility on the status of a published certificate beyond the technical delivery. Finally, Certificate Provider do not have a standard mechanism to send new certificates to the Certificate Consumer when they become available.

Use cases to provide certificates (from who initiates communication, negotiation and data transfer):

- Consumer -> Provider : Company Certificate Request
- Consumer -> Provider : Company Certificate Status (Accepted, Rejected or Received)
- Provider -> Consumer : Company Certificate Push

> - Provider -> Consumer : Company Certificate Available

Reviewer 3: We have added this entry text 6 month ago to enable the notification of a certificate consumer by the certificate provider. I can NOT agree to remove this option!

### 2.1 API Specification

This section introduces the certificate management notification API which is further detailed in the corresponding [OpenAPI specification](assets/openapi-spec.yaml).

#### 2.1.1 API endpoints and resources

##### 2.1.1.1 Company Certificate Request

The Certificate Consumer is requesting a specific certificate from the Certificate Provider.
This request can be sent by the Certificate Consumer continuously, for updates on the request state on Certificate Provider side.

![alt text](images/state-machine-certificate-distributor.svg "Certificate Request API State Machine")

> `POST /companycertificate/request`

Reviewer 3: I was asked to check if the notification mechanismm used in CCM is compliant to CX_0151_Industry Core Basics. The following example is provided in this standard: 1.4 EXAMPLES
Notifcation in JSON format
{
  "header": {
    "messageId": "urn:uuid:f9a97301-a000-44dd-b9d8-78488a40c6bb",
    "context": "IndustryCore-DigitalTwinEventAPI-ConnectToParent:3.0.0",
    "sentDateTime": "2024-07-05T08:13:33.20733Z",
    "senderBpn": "BPNL000000000AAA",
    "receiverBpn": "BPNL000000000ZZZ",
    "expectedResponseBy": "2024-07-08T08:13:33.20733Z",
    "version": "3.0.0"
  },
  "content": {}
}

Does the following request notification is compliant to the above standard?
The "expectedResponseBy": ... is missing? Do we need to add something?

Reviewer 2: Examples in general are non-normative and should not be treated as strict criteria

The `expectedResponseBy` is an optional attribute in the notification API from the CX 0151 Industry Core Basics standard, but also missing completely in the example OpenAPI specification here.

I think it would be a good idea to add it as optional attribute to the example OpenAPI specification, but not to the examples in the standard description itself.

We are adding the following 4 Application Programming Interfaces to remediate these shortcommings:
Today, Certificate Receiver do not have a way to request certificates from a Certificate Distributor. Also, Data Certificate Distributor have no visibility on the status of a published certificate beyond the technical delivery. Finally, Certificate Distributor do not have a standard mechanism to send new certificates to the Certificate Receiver when they become available.

Use cases to provide certificates (from who initiates communication, negotiation and data transfer):

- Certificate Receiver -> Certificate Distributor : Company Certificate Request
- Certificate Distributor -> Certificate Receiver : Company Certificate Push

- Consumer -> Provider : Company Certificate Request
- Consumer -> Provider : Company Certificate Status (Accepted, Rejected or Received)
- Provider -> Consumer : Company Certificate Push

> - Provider -> Consumer : Company Certificate Available

Reviewer 2: ~The use case "Company Certificate Available" is lost here.~

EDIT: This is on purpose.

Reviewer 3: We have introduced the "Company Certificate Available" on purpose in April. I can NOT agree to remove the Use cCse/Application Programming Interface.

#### 2.1.1 API endpoints and resources

##### 2.1.1.1 Company Certificate Request

The Certificate Receiver is requesting a specific certificate from the Certificate Distributor.
This request can be sent by the Certificate Receiver continuously, for updates on the request state on Certificate Distributor side.

> ![alt text](assets/State%20Machine%20Certificate%20Distributor.svg "3-Tier example of a digital twin BOM")

Reviewer 2: Alternative text is not making sense here, copy-paste error?

| 400       | Request malformed.                                                        |
| 500       | Internal Server Error.                                                    |

>**EDC Behavior**:
>At the moment (standard release 25.09), the open-source EDC will always proxy a `500` internal server error when it encounters a `4xx` or `5xx` HTTP response code from the API. 
>This means that in the case of a malformed request, while the API should return a `400` status code, the final EDC response that the consumer receives will be `500`. 
>Until a future EDC update changes this behavior to proxy all status codes without changes, applications will need to be able to deal with this technical reality.

The detailed response bodies for HTTP Code 200 are described in 2.1.1.1.2 and following.
HTTP Status Codes 202, 400 and 500 do not come with a response body.

##### 2.1.1.1.2 HTTP Response Body for HTTP Code 200, Status: IN PROGRESS

Case: Certificate Request Still In Process

```json
{
  "header" : {
    "senderBpn" : "BPNL0000000001AB",
    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",

>     "context" : "CompanyCertificateManagement-CCMAPI-Status:1.0.0",

Reviewer 2: ```suggestion
    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
```

    "sentDateTime" : "2025-05-04T00:00:00-07:00",
    "version" : "3.1.0"
  },
  "content" : {
    "requestStatus":"IN_PROGRESS"
  }
}
```

##### 2.1.1.1.3 HTTP Response Body for HTTP Code 200, Status: COMPLETED

Finished Processing and Certificate Available in EDC.
The content body also provides the documentId of the certificate.
This allows for skipping the EDC catalog search.

```json
{
  "header" : {
    "senderBpn" : "BPNL0000000001AB",
    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
    "messageId" : "3b4edc05-e214-47a1-b0c2-1d831cdd9ba9",
    "receiverBpn" : "BPNL0000000002CD",
    "sentDateTime" : "2025-05-04T00:00:00-07:00",
    "version" : "3.1.0"
  },
  "content": {
   "documentId" : "00000000-0000-0000-0000-000000000001",
   "requestStatus":"COMPLETED"
  }
}
```

##### 2.1.1.1.4 HTTP Response Body for HTTP Code 200, Status: REJECTED

Finished Processing and Certificate Request Rejected.
The request errors and location errors SHOULD contain all encountered problems in detail.
The error message is free text.

```json
{
  "header" : {
    "senderBpn" : "BPNL0000000001AB",
    "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",

>     "context" : "CompanyCertificateManagement-CCMAPI-Status:1.0.0",

Reviewer 2: ```suggestion
    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",
```

    "messageId" : "3b4edc05-e214-47a1-b0c2-1d831cdd9ba9",
    "receiverBpn" : "BPNL0000000002CD",
    "sentDateTime" : "2025-05-04T00:00:00-07:00",
    "version" : "3.1.0"
  },
  "content" : {
    "requestStatus":"REJECTED",
    "requestErrors" : [
      {"message":"We do not process certificates on Sunday"},
      {"message":"Can not provide certicate for requested locations"}
    ],
    "locationErrors" : [
        { "bpn":"BPNS000000000003", "locationErrors": [{"message":"Site BPNS000000000003 is unknown"}]}
    ]
  }
}
```

##### 2.1.1.2 Company Certificate Push

Certificate is pushed by the Certificate Distributor to the Certificate Receiver.
The enclosed BPNs can be a mix of sites and addresses.

> The Certificate Receiver may want to send a subsequent GET or fetch the asset in the catalog.

Reviewer 2: Why would the Certificate Receiver may want to do that? They just received the certificate via the PUSH notification.
The Certificate Receiver may want to send a feedback/status notification afterwards.

Reviewer 4: I think this was a re-wording of the previous sentence and this may have slipped in by accident. I'll change it to "may want to send a subsequent feedback message".

| 202       | Certificate request accepted and in processing.                           |
| 400       | Request malformed.                                                        |
| 500       | Internal Server Error.                                                    |

>**EDC Behavior**:
>At the moment (standard release 25.09), the open-source EDC will always proxy a `500` internal server error when it encounters a `4xx` or `5xx` HTTP response code from the API. 
>This means that in the case of a malformed request, while the API should return a `400` status code, the final EDC response that the consumer receives will be `500`. 
>Until a future EDC update changes this behavior to proxy all status codes without changes, applications will need to be able to deal with this technical reality.

The detailed response bodies for HTTP Code 200 are described in 2.1.1.1.2 and following.
HTTP Status Codes 202, 400 and 500 do not come with a response body.

##### 2.1.1.1.2 HTTP Response Body for HTTP Code 200, Status: IN PROGRESS

Case: Certificate Request Still In Process

```json
{
  "header" : {
    "senderBpn" : "BPNL0000000001AB",

>     "relatedMessageId" : "d9452f24-3bf3-4134-b3eb-68858f1b2362",

Reviewer 2: The example for relatedMessageId is a little bit confusing because there is no message with such an ID anywhere in this or other examples of the standard. As this is an optional field, better leave this out (here and on other similar occasions).
```suggestion
```

Author: I'm more than happy to remove this field but this constitutes a breaking change. Do we know that nobody is relying on it?

Reviewer 5: It should be removed. This is the initial message within the flow which means there cannot be a relatedMessageId.

    "context" : "CompanyCertificateManagement-CCMAPI-Request:1.0.0",

>     "messageId" : "3b4edc05-e214-47a1-b0c2-1d831cdd9ba9",

Reviewer 2: The examples are all reusing the same `messageId` which might indicate that the ID should be reused, but from my understanding there should be a new `messageId` for every individual HTTP request. I would suggest to also use different IDs for the examples.

Author: I set a different UUID for each messageID and relatedMessageID value

    "receiverBpn" : "BPNL0000000002CD",
    "sentDateTime" : "2025-05-04T00:00:00-07:00",
    "version" : "3.1.0"
  },
  "content" : {
    "requestStatus":"IN_PROGRESS"
  }
}
```

##### 2.1.1.1.3 HTTP Response Body for HTTP Code 200, Status: COMPLETED

Finished Processing and Certificate Available in EDC.
The content body also provides the documentId of the certificate.

> This allows for skipping the EDC catalog search.

Reviewer 2: `This allows for skipping the EDC catalog search.`

This is from my understanding a wrong statement because we need the Asset ID, the Offer ID AND the permissions to be able to initiate a EDR token negotiation. Even if the Asset ID would be enough there is no guarantee that this is the same as the document ID - there is no written requirement to use the document ID as Asset ID.

Author: That is true. The documentID is the assetID and merely helps with finding the correct offer in the catalog. I adapted this information accordingly

  }
}
```
The ´senderFeedbackUrl´ specifies, where the Certificate Distributor expects the feedback of the Certificate Receiver.
The expected value **MUST** be a concrete path to the version 1 dataspace protocol endpoint (as specified in the example above),
where a data offer for an asset of type cx-taxo:CCMAPI **MUST** be available for the Certificate Receiver.

>**Push header `senderFeedbackUrl` explanation**: 
>This information is intended as a temporary solution to support the unique identification of multiple feedback endpoints across multiple EDCs belonging to one legal entity. 
>The typical way to implement such differentiation in the Catena-X data space would be to provide additional, distinguishing attributes to the EDC assets to enable an automated search mechanism via the EDC discovery service and EDC catalogs. 
>Since the current changes are implemented as a non-breaking standard patch, the senderFeedbackUrl remains an intermediate solution. 
>A future change is required in that regard, especially when considering the deprecation of the v1 DSP endpoint in favor of an upcoming EDC `.well-known` endpoint that supports multiple DSP versions.

> ##### 2.1.1.3 Company Certificate Feedback

Reviewer 2: Renaming this only here without renaming it on other occasions (e.g. in the `context`) is just introducing more complexity and confusion.
If you plan to rename it, please make sure to do that everywhere.

Reviewer 4: We renamed it, because "Status" would then be confusing with the status of the request api in the pull scenario. We explicitly named it "Feedback' API, because that is what this API is for: providing feedback.

Reviewer 2: Ok, i understand.
All in all a good change but please consider also changing other occasions (like the `context`) to avoid confusion.

Reviewer 4: Good point, I'll change that - that was a copy paste oversight :D

Reviewer: The renaming from `POST /companycertificate/status` to `POST /companycertificate/feedback` would, as I understand it, be a breaking change for the current solution, correct?

`POST /companycertificate/feedback`

> This API is used by the dataspace participant who consumes a certificate to give a feedback to the certificate provider, thus either accepting or rejecting the provided certificate. This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).

Reviewer 2: ```suggestion
This API is used by the Certificate Receiver to give a feedback to the Certificate Distributor, thus either accepting or rejecting the provided certificate. This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).
```

Reviewer 4: I'll also change it then to Certificate Provider and Certificate Consumer, based on todays voting results.

##### 2.1.1.3.1 Company Certificate Feedback: Accepted

> Certificate is accepted. Document UUID should match the incoming message. The enclosed BPNs can be a mix of sites and addresses

Reviewer 2: ```suggestion
Certificate is accepted. Document UUID should match the one from the incoming message. The enclosed BPNs can be a mix of sites and addresses
```

  }
}
```
The ´senderFeedbackUrl´ specifies, where the Certificate Provider expects the feedback of the Certificate Consumer.

> The expected value **MUST** be a concrete path to the version 1 dataspace protocol endpoint (as specified in the example above),

Reviewer 2: nit: This here is a normative MUST requirement which is referencing a non-normative example.

Author: I removed the remark in the paranthesis

where a data offer for an asset of type cx-taxo:CCMAPI **MUST** be available for the Certificate Consumer.

> 

Reviewer 2: nit: A hint that the Push-Notification is the only notification where the attribute `documentID` is spelled with a capital D at the end (due to the spelling in the aspect model) might be good here. This is prone to errors when implementing.

Author: I added such a hint

>**Push header `senderFeedbackUrl` explanation**: 
>This information is intended as a temporary solution to support the unique identification of multiple feedback endpoints across multiple EDCs belonging to one legal entity. 
>The typical way to implement such differentiation in the Catena-X data space would be to provide additional, distinguishing attributes to the EDC assets to enable an automated search mechanism via the EDC discovery service and EDC catalogs. 
>Since the current changes are implemented as a non-breaking standard patch, the senderFeedbackUrl remains an intermediate solution. 
>A future change is required in that regard, especially when considering the deprecation of the v1 DSP endpoint in favor of an upcoming EDC `.well-known` endpoint that supports multiple DSP versions.

##### 2.1.1.3 Company Certificate Feedback

`POST /companycertificate/feedback`

This API is used by the Certificate Consumer to give a feedback to the Certificate Provider, thus either accepting or rejecting the provided certificate.
This is regardless of whether the certificate was [pulled](#2111-company-certificate-request) or [pushed](#2112-company-certificate-push).

> ##### 2.1.1.3.1 Company Certificate Feedback: Accepted

Reviewer 2: nit: Ordering is strange here, i would expect the `Received` status to be the first one because the `Received` status only makes sense to return before `Accepted` or `Rejected`.

Author: Agreed, reordered

Certificate is accepted. 

> The documentId should match the documentId from the incoming message. 

Reviewer 2: `The documentId should match the documentId from the incoming message.`
1. `should` not written in all-capital-letters, is this intended to be non-normative?
2. There might not be an incoming message if the certificate was pulled without previous Request-notification.

Author: I think this should be a MUST and worded it now like that. I also rephrased what the expected value of the documentId should be be.

> The enclosed BPNs can be a mix of sites and addresses.

Reviewer 2: ```suggestion
The `locationBpns` can be a mix of sites and addresses.
```

Author: Done

|----------------|------------------------------------------|-------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateDistributorApi | 3.0         | Offers *Certificate Distributor API* for the Certificate Receiver to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateReceiverApi    | 3.0         | Offers *Certificate Receiver API* for the Certificate Distributor to [push](#2112-company-certificate-push) certificates to the Certificate Receiver.                                                             |

*Example Certificate Distributor API:*
```json
{

>   "@id": "CCMAPI",

Reviewer 2: This is misleading as there can't be two assets with the same assetId on one connector.
How does the setup look if we want to have both APIs configured on one connector?

Reviewer 4: Both APIs share the same asset, so you would not have two assets, but only one. The standard then says APIs must be able to return a 500 error; so if one of the APIs is not available (since only solution providers must provide both APIs) it would need throw a 500 when calling that API.

Reviewer 4: never mind - I thought you referred to the case of the Certificate Distributor

regarding the ```"@id": "CCMAPI"``` -> this should be a (dataspace participant individual / random) uuid anyways, as this could be misleading otherwise. 
Good point, we should also comment that.

Finding / distinguishing the EDC assets would be done by this:
```json
"dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi"
},
```
and
```json
"dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateReceiverApi"
},
```
respectively.

    },
    "dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi"
    },
    "dct:description": "Offers API for the Certificate Receiver to request certificates and send feedback for provided certificates.",
    "cx-common:version": "3.0"
  },
  "dataAddress": {},
  "@context": {
    "dct": "http://purl.org/dc/terms/",
    "cx-taxo": "https://w3id.org/catenax/taxonomy#",
    "cx-common": "https://w3id.org/catenax/ontology/common#"
  }
}
```

*Example Certificate Receiver API:*
```json
{
  "@id": "CCMAPI",
  "@type": "Asset",
  "properties": {
    "dct:type": {
      "@id": "cx-taxo:CCMAPI"
    },
    "dct:subject": {
      "@id": "cx-taxo:CompanyCertificateManagementCertificateReceiverApi"
    },
    "dct:description": " Offers API for the Certificate Distributor to push certificates to the Certificate Receiver.",
    "cx-common:version": "3.0"
  },
  "dataAddress": {},
  "@context": {
    "dct": "http://purl.org/dc/terms/",
    "cx-taxo": "https://w3id.org/catenax/taxonomy#",
    "cx-common": "https://w3id.org/catenax/ontology/common#"
  }
}
```

#### 2.1.5 MESSAGE FLOW EXPECTATIONS

Certificate Distributor & Certificate Receiver:

> - Certificate Distributor MUST expose company certificates in their catalog when using the pull mechanism.

Reviewer 6: @FEILSTE - Concerning "pull mechanism" are we also referring to the "standard" EDC catalog search and the registration of individual certificates within the catalog?
If yes, then I believe we need to provide the EDC asset structure as well. This will ensure that the Certificate Distributor understands how to correctly register the asset in EDC.
Here is an example:
```json
{
    "@id": "CompanyCertficate",
    "@type": "Asset",
    "properties": {
        "dct:certificateType": {
            "@id": "cx-taxo:ISO9001" 
        },
         "dct:enclosedSites": [
            {
                "@id": "cx-taxo:BPNS000000000001"
            },
            {
                "@id": "cx-taxo:BPNS000000000002"
            },
            {
                "@id": "cx-taxo:BPNA000000000001"
            }
         ]
        "dct:subject": {
            "@id": "cx-taxo:CompanyCertificate"
        },
        "dct:description": "Business Partner Company Certificate",
        "cx-common:version": "3.0",
    },
    "dataAddress": {
      "@type": "DataAddress",
      "type": "HttpData",
      "baseUrl": "https://backend-base-url/certificate-management-api-base-path",
      "proxyQueryParams": "true",
      "proxyPath": "false",
      "proxyMethod": "false",
      "proxyBody": "false"
    },
    "@context": {
        "dct": "http://purl.org/dc/terms/",
        "cx-taxo": "https://w3id.org/catenax/taxonomy#",
        "cx-common": "https://w3id.org/catenax/ontology/common#"
    }
}
```

Reviewer 4: @Cvetanka-ZF Good point, we'll add that - thank you :)

Reviewer 4: I added it in the latest change -> feel free to provide feedback regarding the example and the normative text beforehand that defines how the certificateType and enclosedSites are specified.

Reviewer 5: > @FEILSTE - Concerning "pull mechanism" are we also referring to the "standard" EDC catalog search and the registration of individual certificates within the catalog? If yes, then I believe we need to provide the EDC asset structure as well. This will ensure that the Certificate Distributor understands how to correctly register the asset in EDC. Here is an example:
> 
> ```json
> {
>     "@id": "CompanyCertficate",
>     "@type": "Asset",
>     "properties": {
>         "dct:certificateType": {
>             "@id": "cx-taxo:ISO9001" 
>         },
>          "dct:enclosedSites": [
>             {
>                 "@id": "cx-taxo:BPNS000000000001"
>             },
>             {
>                 "@id": "cx-taxo:BPNS000000000002"
>             },
>             {
>                 "@id": "cx-taxo:BPNA000000000001"
>             }
>          ]
>         "dct:subject": {
>             "@id": "cx-taxo:CompanyCertificate"
>         },
>         "dct:description": "Business Partner Company Certificate",
>         "cx-common:version": "3.0",
>     },
>     "dataAddress": {
>       "@type": "DataAddress",
>       "type": "HttpData",
>       "baseUrl": "https://backend-base-url/certificate-management-api-base-path",
>       "proxyQueryParams": "true",
>       "proxyPath": "false",
>       "proxyMethod": "false",
>       "proxyBody": "false"
>     },
>     "@context": {
>         "dct": "http://purl.org/dc/terms/",
>         "cx-taxo": "https://w3id.org/catenax/taxonomy#",
>         "cx-common": "https://w3id.org/catenax/ontology/common#"
>     }
> }
> ```
this is what the asset properties look like in T-Systems implementation:

"semanticId": "io.catenax.business_partner_certificate",
                "dct:type": {
                    "@id": "https://w3id.org/catenax/taxonomy#Submodel"
                },
                "aas:semanticId": {
                    "@id": "urn:samm:io.catenax.business_partner_certificate:3.1.0#BusinessPartnerCertificate"
                },
                "name": "TestCertificate - IATF16949 - RegNo 55558",
                "id": "a6c9a5d4-1dbf-42b6-8b6d-2751d4d13cb7",
                "https://w3id.org/catenax/ontology/common#version": "3.0",
                "assettype": "Submodel",
                "contenttype": "application/json"

Reviewer 4: This implementation is - as of right now - not supported by the standard (as it requires asset discovery based on searching the digital twin registry - the current search mechanism is based on EDC asset properties); it is my understanding that this standard never intended to use digital twins for the certificates, but rather an API based data-exchange. If this is to change then the standard would have to be re-written in agreeance with all expert group participants (as it would require quite a large change, possible breaking - which is only doable in a major release). But as I was not part of this standard development process before the 25.09 release, someone else should confirm this (@nicoprow / @Cvetanka-ZF one of you or someone you know?)

#### 2.1.3 Available data types

The API **MUST** use JSON formatted data transmitted over HTTPS.

#### 2.1.4  Data asset structure

> *This section is normative*

> The HTTP endpoints introduced in chapter [2.1.1 API endpoints and resources](#211-API-endpoints-and-resources) **MUST NOT** be called from a business partner directly. Rather, it **MUST** be called via a connector communication.

Reviewer 2: ```suggestion
The HTTP endpoints introduced in chapter [2.1.1 API endpoints and resources](#211-API-endpoints-and-resources) **MUST NOT** be called from a business partner directly. Rather, it **MUST** be called with any CX-0018 compliant connector via dataspace protocol.
```

Author: Done

Therefore, the Certificate Provider **MUST** offer an asset to expose an API for Certificate Consumer in the connector catalog.
In turn, the Certificate Consumer **MAY** offer an asset to expose an API for the Certificate Provider in the connector catalog to push certificates to.

The property [[type]](http://purl.org/dc/terms/type) **MUST** reference the name of the certificate management API as defined in the Catena-X taxonomy published under [[taxonomy]](https://w3id.org/catenax/taxonomy).

| **Type**       | **Subject**                              | **Version** | **Description**                                                                                                                                                                                                |
|----------------|------------------------------------------|-------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | cx-taxo:CCMAPI | cx-taxo:CompanyCertificateDistributorApi | 3.0         | Offers *Certificate Provider API* for the Certificate Consumer to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |

Reviewer 2: Why is a different naming introduced here?
In my opinion this just adds more complexity, sticking to the already used terms would be better.
```suggestion
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateProviderApi | 3.0         | Offers *Certificate Provider API* for the Certificate Consumer to [request](#2111-company-certificate-request) certificates and send [feedback](#2113-company-certificate-feedback) for provided certificates. |
```

Author: Leftover from previous terms. I changed it to `Provider`

> | cx-taxo:CCMAPI | cx-taxo:CompanyCertificateReceiverApi    | 3.0         | Offers *Certificate Consumer API* for the Certificate Provider to [push](#2112-company-certificate-push) certificates to the Certificate Consumer.                                                             |

Reviewer 2: ```suggestion
| cx-taxo:CCMAPI | cx-taxo:CompanyCertificateConsumerApi    | 3.0         | Offers *Certificate Consumer API* for the Certificate Provider to [push](#2112-company-certificate-push) certificates to the Certificate Consumer.                                                             |
```

Author: Leftover from previous terms, I changed it to `Consumer`

There **MUST** only be one unique asset per API (subject and version) across all connectors of one BPNL.

*Example*: it is possible to have these assets available next to one-another:
- ```{"dct:subject"."@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi", "cx-common:version": "3.0"}```, and 
- ```{"dct:subject"."@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi", "cx-common:version": "2.0"}```, as well as
- ```{"dct:subject"."@id": "cx-taxo:CompanyCertificateManagementCertificateReceiverApi", "cx-common:version": "3.0"}```,

since they either differ in the value of the version or the subject. But it would not be possible to have two of the same subject and the same version.

*Example that is not allowed:*
- ```{"dct:subject"."@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi", "cx-common:version": "3.0"}```
- ```{"dct:subject"."@id": "cx-taxo:CompanyCertificateManagementCertificateDistributorApi", "cx-common:version": "3.0"}```

It doesn't matter if the assets are offered in one or in different connectors, as long as they belong to the same BPNL this is not allowed.

> *Example Certificate Provider API:*

Reviewer 2: nit: Introducing sub-sections for each individual asset example would make it easier to recognize which text-block is belonging to which example.

Author: Restructured the section a bit and added subsections for clarity

    "dct:description": " Offers API for the Certificate Provider to push certificates to the Certificate Consumer.",
    "cx-common:version": "3.0"
  },
  "dataAddress": {},
  "@context": {
    "dct": "http://purl.org/dc/terms/",
    "cx-taxo": "https://w3id.org/catenax/taxonomy#",
    "cx-common": "https://w3id.org/catenax/ontology/common#"
  }
}
```

The certificate assets **MUST** be created by the Certificate Provider in their connector catalog to be consumed by the Certificate Consumer.
The property certificateType **MUST** reference the type of the certificate as defined in [3.2.2 Certificate Type](#322-certificate-type) (without spaces in the certificate type).
The property enclosedSites **MUST** contain all BPNSs for which the certificate is valid.
The subject **MUST** reference cx-taxo:CompanyCertificate.
Additionally, the assets **MUST** contain the type ```cx-taxo:Submodel``` and the semanticId specified in [3.1.2 Business Partner Company Certificate Submodel](#312-business-partner-company-certificate-submodel).

*Example Certificate EDC Asset:*
```json
{
    "@id": "d195fa2f-e6bc-4cd6-94d4-2bb76e4bb548",
    "@type": "Asset",
    "properties": {
        "dct:type": {

>             "@id": "https://w3id.org/catenax/taxonomy#Submodel"

Reviewer 2: ```suggestion
            "@id": "cx-taxo:Submodel"
```

Author: Done

        },
        "aas:semanticId": {
            "@id": "urn:samm:io.catenax.business_partner_certificate:3.1.0#BusinessPartnerCertificate"
        },
        "dct:certificateType": {
            "@id": "cx-taxo:ISO9001" 
        },
         "dct:enclosedSites": [
            {
                "@id": "cx-taxo:BPNS000000000001"
            },
            {
                "@id": "cx-taxo:BPNS000000000002"
            },
            {
                "@id": "cx-taxo:BPNA000000000001"
            }
         ],
        "dct:subject": {
            "@id": "cx-taxo:CompanyCertificate"
        },
        "dct:description": "Business Partner Company Certificate",
        "cx-common:version": "3.0"
    },
    "dataAddress": {
      "@type": "DataAddress",
      "type": "HttpData",
      "baseUrl": "https://backend-base-url/certificate-management-api-base-path",

>       "proxyQueryParams": "true",

Reviewer 2: Why is `proxyQueryParams` set to true? AFAIK we don't utilize them when pulling certificates.

```suggestion
      "proxyQueryParams": "false",
```

Author: As far as I know query parameters are not needed here. Changed to false

}
```

> The **certificate assets** are identified by the combination of `dct:subject`, `dct:certificateType` and `dct:enclosedSites`.
> When searching the EDC catalog for a specific asset for which the assetId is unkown, make use of those properties in the catalog filter.

#### 2.1.5 MESSAGE FLOW EXPECTATIONS

- Data Provider MUST expose company certificates in their catalog when using pull mechanism.
- Data Provider MUST set the correct access policy on the certificate asset to allow consumption by Consumer(s) when using pull mechanism.

> - Data Consumer MUST set the correct access policy on the notification policy to allow the exchange of data by provider when using push mechanism.

Reviewer 3: Why does this statement was removed?
Assumption: When the push mechanism is used, the certificate provider needs access to the EDC of the certificate consumer to provide the certificate data.
What is the difference between "setting an access policy on the notification policy" and setting an "access policy on the offer" by the data consumer?

Reviewer 3: I have got the answer from the Data Sovereignity EG.  "access policies on the offer" is the correct formulation.
However this leads to a required correction in chapter 2.1.6
Current text: 
"2.1.6 Usage Policy
The “Connector Asset for Certificate Notifications” included in the EDC of a data consumer MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version. That includes the following usage purpose:
•	cx.ccm.base:1: The exchanged business partner certificates are used for the purpose of verification of the existence of a certification. As a purpose-specific requirement, the duration of (i) contract, (ii) data provision and (iii) usage right(s) as a default are all specified as 1 year.
Additional more general usage policies MAY be included, but all the usage policies MUST contain the above usage purpose."

We have to agree to the new text. The following changes are required:
1. The "Connector Asset for Certificate Notification" has to removed and replaces with "EDC" respective "EDC offer". I do not know the correct term for the data structure in the EDC that contains the usage purpose and the contract reference.
2. "MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version. " is a contradiction to the statement that "Version 1.0" of the usage purpose should apply made in the changes above. The contradiction needs to be removed.

> - Certificate Provider **MUST** expose company certificates in their catalog when using the pull mechanism.

Reviewer 3: What does "expose company certificates" means?

Reviewer 2: In this context this means that in order for a receiver to consume certificates via the PULL-mechanism from the provider, the provider must ensure that for every consumable certificate there has to be a corresponding asset in their catalog.

- Certificate Provider **MUST** set the correct access policy on the certificate offer to allow consumption by Consumer(s) when using the pull mechanism.

> - Certificate Consumer **MAY** implement the [push endpoint](#2112-company-certificate-push) for the Certificate Provider to push certificates to, but **MUST** set the correct access policy on the offer, when choosing to do so.

Reviewer 3: It is mandatory that certificate consumers"... **MUST* set the correct access policy **and also the correct bilateral contract reference**..." Please change the corresponding sentence above.

Reviewer 2: Please take a look at line 635:

> The constraint `{ "leftOperand": "cx-policy:ContractReference" }` **MUST** be included only if such a bilateral framework contract exists.

Is this what you are referencing?

##### 2.1.5.3 PUSH notification followed by PULL mechanism

After the data provider has created a Certificate Asset with corresponding contract definition in the EDC Catalog the data provider sends a Certificate Available Notification to the data consumer. The data consumer uses the above described PULL mechanism to get the certificate data. This reduces the data consumers need for active checks for missing certificates or certificate updates and enables access to the latest certificate data.
It begins with the provider creating a Certificate Asset with corresponding contract definition in the EDC Catalog. 
The Consumer searches the catalog using specific filters, initiates a contract negotiation, and retrieves the Endpoint Data Reference (EDR). 
The Data Plane then facilitates secure data transfer, allowing the consumer to pull the certificate. 
Once retrieved, the Backend Certificate Consumer processes the certificate and sends a Feedback Message to confirm the status.

##### 2.1.6 Usage Policy

> The “Connector Asset for Certificate Notifications” included in the EDC of a data consumer MUST contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version.

Reviewer 3: What is the benefit of the reference to version 1.0 instead the reference to "the latest version"?

Reviewer 2: Quoting my comment from a previous review:
> nit: Might not be a good idea to force the latest version as newer policy versions might not be compatible. Better define a fixed version, the current latest one.

The left operand "leftOperand": "cx-policy:ContractReference" MUST be included only if such a bilateral framework contract exists.

```json

>             {

Reviewer 2: ```suggestion
{
```

}
```

>**`@id` explanation**:
> Each of the above examples has a random example uuid. 
> Every dataspace participant may use their individual random uuid.

> The **API assets** are identified by the combination of `dct:subject` and `cx-common:version`.
> When searching the EDC catalog for a specific API, make use of those properties in the catalog filter.

> The **certificate assets** are identified by the combination of `dct:subject`, `dct:certificateType` and `dct:enclosedSites`.
> When searching the EDC catalog for a specific asset for which the assetId is unkown, make use of those properties in the catalog filter.

#### 2.1.5 MESSAGE FLOW EXPECTATIONS

Certificate Provider & Certificate Consumer:
- Certificate Provider **MUST** expose company certificates in their catalog when using the pull mechanism.
- Certificate Provider **MUST** set the correct access policy on the certificate offer to allow consumption by Consumer(s) when using the pull mechanism.

> 

Reviewer 2: What about the Certificate Provider API? Is this a **MUST** requirement?

- Business Application Provider **MUST** implement all features of the Certificate Consumer API, for supporting the push mechanism.
- Business Application Provider **MUST** offer the push mechanism option to the Certificate Consumer application user, if the Certificate Consumer supports the push mechanism.

##### 2.1.5.1 PUSH Mechanism

![PUSH Scenarios](images/certificate-push.svg)

The Certificate PUSH Diagram describes the secure transmission of certificates from a Backend Certificate Provider to a Backend Certificate Consumer via EDC (Eclipse Data Connector) components.
The process starts with a contract agreement for a Notification Asset, followed by the provider pushing the certificate to the provided endpoint in the asset.  The certificate is then processed by the Backend Certificate Consumer, which finalizes the workflow by generating a feedback message which is pushed to the provider.

##### 2.1.5.2 PULL Mechanism

![PULL Scenarios](images/certificate-pull.svg)

The Certificate PULL Diagram describes the process of Consumer retrieving a certificate from a Provider via an EDC.

> It begins with the provider creating a Certificate Asset with corresponding contract definition in the EDC Catalog. The Consumer searches the catalog using specific filters, initiates a contract negotiation, and retrieves the Endpoint Data Reference (EDR). The Data Plane then facilitates secure data transfer, allowing the consumer to pull the certificate. Once retrieved, the Backend Certificate Consumer processes the certificate and sends a Feedback Message to confirm the status.

Reviewer 2: The whole sub-section `2.1.5.3 PUSH notification followed by PULL mechanism` doesn't make any sense to me, both mechanisms are for consuming certificates, therefore you can use one OR the other, but one following the other doesn't make sense to me (for normal cases, one potential use case could be an update of a certificate).

Also, there are no `Certificate Available Notification`s anymore which are mentioned there.

Author: Looks to me like a legacy section we forgot to adapt. I also didn't see relevance now and removed it.

##### 2.1.6 Usage Policy

> All assets included in the EDC of a dataspace participant (APIs and Certificates) **MUST** contain a usage policy that includes the Catena-X Data Exchange Governance document in the latest version.

Reviewer 2: nit: Might not be a good idea to force the latest version as newer policy versions might not be compatible. Better define a fixed version, the current latest one.

Author: Makes sense to me. I added 1.0 version as this is the current one.

The use-case introduces the following usage purpose:

- **`cx.ccm.base:1:`** *The exchanged business partner certificates are used for the purpose of verification and validation of the existence of a certification.*

Additional more general usage policies **MAY** be included, 
but all the usage policies **MUST** contain the above usage purpose.

*Minimal example of a usage policy without contract reference:*
``` json
{

>   "@context": [

Reviewer 2: The `@context` example here looks strange to me, i haven't seen it like this before. I'm not sure if this can be an array.

Author: This is actually legit and working

}
```

> The left operand "leftOperand": "cx-policy:ContractReference" **MUST** be included only if such a bilateral framework contract exists.

Reviewer 2: ```suggestion
The constraint `{ "leftOperand": "cx-policy:ContractReference" }` **MUST** be included only if such a bilateral framework contract exists.
```

Author: Applied the changes

```json
{
  "leftOperand": "cx-policy:ContractReference",
  "operator": "eq",
  "rightOperand": "x12345"
}
```

> ## 3 ASPECT MODELS

Reviewer 2: nit: Add links to mentions of `[CX-0003:1.1]`

Author: Added the links
